File: go/tests/sports_board_test.go
// sports_board_test.go
//
// Verifies the CircleAI.Sports port (sports_board.go): logging + newest-first
// history with limit, weekly distance totals, best-time lookup, and session
// schedule/complete/upcoming.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestSports_HistoryAndWeek(t *testing.T) {
	b := circleai.NewInMemorySportsBoard()
	// A Wednesday so the week window (Sunday..) is unambiguous.
	now := time.Date(2026, 7, 8, 12, 0, 0, 0, time.UTC)
	b.Log(circleai.SportsActivity{ActivityId: "a1", UserId: "u1", Kind: circleai.DistanceKindRun, DistanceKm: 5, Duration: 25 * time.Minute, AtUtc: now})
	b.Log(circleai.SportsActivity{ActivityId: "a2", UserId: "u1", Kind: circleai.DistanceKindRun, DistanceKm: 10, Duration: 55 * time.Minute, AtUtc: now.Add(24 * time.Hour)})
	b.Log(circleai.SportsActivity{ActivityId: "a3", UserId: "u1", Kind: circleai.DistanceKindRun, DistanceKm: 3, Duration: 15 * time.Minute, AtUtc: now.Add(-14 * 24 * time.Hour)}) // last fortnight
	b.Log(circleai.SportsActivity{ActivityId: "b1", UserId: "u2", Kind: circleai.DistanceKindRun, DistanceKm: 99, Duration: time.Hour, AtUtc: now})

	hist := b.History("u1", 50)
	if len(hist) != 3 || hist[0].ActivityId != "a2" {
		t.Fatalf("history newest-first failed: %+v", hist)
	}
	if lim := b.History("u1", 1); len(lim) != 1 || lim[0].ActivityId != "a2" {
		t.Fatalf("history limit failed: %+v", lim)
	}
	if km := b.TotalKmThisWeek("u1", circleai.DistanceKindRun, now); km != 15 {
		t.Fatalf("weekly km = %v, want 15", km)
	}
}

func TestSports_BestAndSessions(t *testing.T) {
	b := circleai.NewInMemorySportsBoard()
	now := time.Date(2026, 7, 8, 12, 0, 0, 0, time.UTC)
	b.Log(circleai.SportsActivity{ActivityId: "a1", UserId: "u1", Kind: circleai.DistanceKindRun, DistanceKm: 5, Duration: 30 * time.Minute, AtUtc: now})
	b.Log(circleai.SportsActivity{ActivityId: "a2", UserId: "u1", Kind: circleai.DistanceKindRun, DistanceKm: 5, Duration: 22 * time.Minute, AtUtc: now})

	best, ok := b.Best("u1", circleai.DistanceKindRun, 5)
	if !ok || best.Time != 22*time.Minute {
		t.Fatalf("best = %+v ok=%v, want 22m", best, ok)
	}
	if _, ok := b.Best("u1", circleai.DistanceKindSwim, 5); ok {
		t.Fatalf("best for absent kind must be false")
	}

	future := time.Now().UTC().Add(48 * time.Hour)
	b.Schedule(circleai.TrainingSession{SessionId: "s1", UserId: "u1", Plan: "Intervals", ScheduledUtc: future})
	if up := b.Upcoming("u1"); len(up) != 1 || up[0].SessionId != "s1" {
		t.Fatalf("upcoming failed: %+v", up)
	}
	if err := b.Complete("s1"); err != nil {
		t.Fatalf("complete: %v", err)
	}
	if up := b.Upcoming("u1"); len(up) != 0 {
		t.Fatalf("completed session should not be upcoming: %+v", up)
	}
	if err := b.Complete("ghost"); err == nil {
		t.Fatalf("completing unknown session must error")
	}
}

File: typescript/tests/sports_board.test.ts
// sports_board.test.ts
// Verifies the CircleAI.Sports port: activity log + history ordering, weekly km,
// best-over-distance, scheduling + completion + upcoming.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemorySportsBoard,
  SportsDomainContext,
  DistanceKind,
  activity,
  trainingSession,
} from "../src/sports/index";

describe("InMemorySportsBoard", () => {
  it("logs activities and returns history newest-first, capped by limit", () => {
    const b = new InMemorySportsBoard();
    b.log(activity("a1", "u1", DistanceKind.Run, 5, 1_500_000, new Date("2026-01-01T06:00:00Z")));
    b.log(activity("a2", "u1", DistanceKind.Run, 10, 3_000_000, new Date("2026-01-03T06:00:00Z")));
    b.log(activity("a3", "u1", DistanceKind.Run, 3, 900_000, new Date("2026-01-02T06:00:00Z")));
    b.log(activity("a4", "u2", DistanceKind.Run, 8, 2_400_000, new Date("2026-01-04T06:00:00Z")));
    assert.deepEqual(
      b.history("u1").map((a) => a.activityId),
      ["a2", "a3", "a1"],
    );
    assert.deepEqual(
      b.history("u1", 2).map((a) => a.activityId),
      ["a2", "a3"],
    );
  });

  it("history throws on non-positive limit", () => {
    const b = new InMemorySportsBoard();
    assert.throws(() => b.history("u1", 0), /limit/);
  });

  it("sums this week's km from the Sunday week-start", () => {
    const b = new InMemorySportsBoard();
    // now = Wed 2026-01-07. Week start = Sun 2026-01-04 00:00Z.
    const now = new Date("2026-01-07T12:00:00Z");
    b.log(activity("a1", "u1", DistanceKind.Run, 5, 1, new Date("2026-01-03T23:59:59Z"))); // before week start
    b.log(activity("a2", "u1", DistanceKind.Run, 10, 1, new Date("2026-01-04T00:00:00Z"))); // exactly week start
    b.log(activity("a3", "u1", DistanceKind.Run, 7, 1, new Date("2026-01-06T09:00:00Z")));
    b.log(activity("a4", "u1", DistanceKind.Bike, 100, 1, new Date("2026-01-06T09:00:00Z"))); // wrong kind
    assert.equal(b.totalKmThisWeek("u1", DistanceKind.Run, now), 17);
    assert.equal(b.totalKmThisWeek("u1", DistanceKind.Bike, now), 100);
  });

  it("best returns the fastest qualifying activity, projected to the requested distance", () => {
    const b = new InMemorySportsBoard();
    b.log(activity("a1", "u1", DistanceKind.Run, 10, 3_000_000, new Date("2026-01-01T06:00:00Z")));
    b.log(activity("a2", "u1", DistanceKind.Run, 12, 2_500_000, new Date("2026-01-02T06:00:00Z")));
    b.log(activity("a3", "u1", DistanceKind.Run, 5, 900_000, new Date("2026-01-03T06:00:00Z"))); // too short
    const best = b.best("u1", DistanceKind.Run, 10);
    assert.ok(best);
    assert.equal(best?.distanceKm, 10);
    assert.equal(best?.timeMs, 2_500_000);
    assert.equal(best?.achievedUtc.toISOString(), "2026-01-02T06:00:00.000Z");
    assert.equal(b.best("u1", DistanceKind.Run, 100), undefined);
  });

  it("schedules sessions, completes them, and lists only future incomplete ones", () => {
    const b = new InMemorySportsBoard();
    const future = new Date(Date.now() + 86_400_000);
    const past = new Date(Date.now() - 86_400_000);
    b.schedule(trainingSession("s1", "u1", "Long run", future, false));
    b.schedule(trainingSession("s2", "u1", "Tempo", new Date(Date.now() + 172_800_000), false));
    b.schedule(trainingSession("s3", "u1", "Old", past, false));
    b.complete("s2");
    assert.deepEqual(
      b.upcoming("u1").map((s) => s.sessionId),
      ["s1"],
    );
  });

  it("complete throws on unknown session", () => {
    const b = new InMemorySportsBoard();
    assert.throws(() => b.complete("ghost"), /Unknown session ghost/);
  });

  it("DistanceKind values match the C# enum ordinals", () => {
    assert.equal(DistanceKind.Run, 0);
    assert.equal(DistanceKind.Row, 4);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(SportsDomainContext.systemPromptSnippet.includes("[DOMAIN: Sports]"));
    assert.deepEqual(SportsDomainContext.complianceFlags, ["WADA", "SASCOC", "Sport_Recreation_SA", "POPIA"]);
    assert.deepEqual(SportsDomainContext.suggestedTools, [
      "performance_tracker",
      "analytics",
      "schedule_manager",
      "document_editor",
    ]);
  });
});
